### PR TITLE
eeprom_stm32: implement wear leveling

### PR DESCRIPTION
**Pull request by xyzz**
This is my take on #7803 implementing flash wear leveling for stm32 emulated eeprom (but without stm32f4 support - it should be added at a later point).

## Description

Currently eeprom emulation is mostly unusable on stm32 due to the fact that on every byte write it performs a full page erase. For example, if you try to set up a macro in via that is 500 characters, qmk would perform 500 page erases which both takes forever and needlessly wears down flash memory.

This change implements flash wear leveling using essentially the same algorithm as in #7803:
- initially, eeprom is assumed to be all-zero
- whenever a write is performed, a change is appended to the "write log" of the format (uint16_t address, uint8_t value). Every change takes 4 bytes, padded with a single 0xFF
- during firmware initialization the write log is parsed to create global array DataBuf which contains current state of eeprom
- if we run out of flash space the entire flash is erased and then reinitialized based on DataBuf contents

Due to the write log implementation, the minimum amount of flash memory required to store 1024 bytes of emulated eeprom is 4096, however I am using 8192 instead so that when eeprom is full, we can still perform writes without having to erase the entire memory region every time.

I have tested these changes with an F103 blue pill board, prior to the changes setting a long macro (900 characters) in via is pretty much impossible (I had to force power off the board), after the changes it takes a couple seconds.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by zvecr** [comment]
Awesome! Splitting out the wear leveling is a good way to break up the previous PR. Balancing the decrease in eeprom vs potential increase in flash usage is something that will have to be considered. 

Also as this would need `EECONFIG_MAGIC_NUMBER` to be updated, this should target `develop`.

**Comment by xyzz** [comment]
Okay it's ready for review now

**Review by tzarc — APPROVED** [review]

**Review by tzarc — CHANGES_REQUESTED** [review]
One change required, please decrease the value of `EECONFIG_MAGIC_NUMBER` by 1, as per @zvecr's suggestion. Rest looks good!

**Review by tzarc — APPROVED** [review]

**Comment by tzarc** [comment]
Sorry about that, GitHub decided to delete the `develop` branch from upstream when we merged it, despite being told otherwise. Reopened.

**Comment by drashna** [comment]
This has some merge conflicts still, could you fix that?

**Comment by xyzz** [comment]
ok, i rebased on latest develop

**Comment by tzarc** [comment]
On further investigation, there are some concerns regarding the validity of this implementation for certain cases.

### STM32F042 builds

* These have 32kB flash
* Under the current model use up 2 x 1kB pages = 2kB for emulated eeprom
* Under the new model, they'd use 8 x 1kB pages = 8kB for emulated eeprom
* Their use of standard layouts provides no guarantee that user keymaps under `layouts/` are in use, rather than relying on the presence of a keymap directory associated with the board itself
* Most keymaps for all F042 builds max out at 23kB flash used, so claiming 8kB for EEPROM would normally _just_ work, but...
* As a "worst case" scenario, building `akegata_denki/device_one:unxmaal` results in a binary of 31,170 bytes -- which would already invalidate the current implementation on `master`. The board itself doesn't have a VIA keymap however, so I'm unsure if this is to be taken into account.

### `DYNAMIC_KEYMAP_EEPROM_MAX_ADDR` extensions

For the boards that increase the value of `DYNAMIC_KEYMAP_EEPROM_MAX_ADDR`, the following boards use emulated STM32 EEPROM:

* `ramonimbao/squishy65` => 2047, STM32F072
* `ramonimbao/wete` => 2047, STM32F072

I'm not sure if there are any other mechanisms in use in order to extend the use of EEPROM -- they've usually been associated with usage of VIA.

For the short term, I think it might be best if we provide the ability to select whether or not wear leveling is available for use. We can default to __off__ for F042 by default, and explicitly disable it for the above two boards.

Alternatively, if we can get confirmation from:

* @ramonimbao -- Is it critical that your two boards expose 2kB of EEPROM to users? Can we decrease the EEPROM usage on your two keyboards above in order to pick up wear leveling on your boards?
* @unxmaal -- Can you confirm that you don't own any of the following keyboards: `akegata_denki/device_one` `chavdai40/rev1` `chavdai40/rev2` `peiorisboards/ixora` `polilla/rev1` `vinta`

...then I think we're okay to proceed with the implementation as-is.

**Comment by ramonimbao** [comment]
> Is it critical that your two boards expose 2kB of EEPROM to users? Can we decrease the EEPROM usage on your two keyboards above in order to pick up wear leveling on your boards? 

I'm fine with this if it works. Though before, I had to increase the EEPROM usage because for some reason the board would not compile with the default settings.

**Comment by tzarc** [comment]
> I'm fine with this if it works. Though before, I had to increase the EEPROM usage because for some reason the board would not compile with the default settings.

Finally had a chance to circle back around to this -- `wete` does not compile as its keymap is too large for 4 layers in VIA.
In order to progress wear leveling, we'd need to decrease the number of available layers for VIA to use to 3, instead of increasing the EEPROM usage:
```diff
diff --git a/keyboards/ramonimbao/wete/keymaps/via/config.h b/keyboards/ramonimbao/wete/keymaps/via/config.h
index 4dadf6a16..1f4e07e39 100644
--- a/keyboards/ramonimbao/wete/keymaps/via/config.h
+++ b/keyboards/ramonimbao/wete/keymaps/via/config.h
@@ -1 +1 @@
-#define DYNAMIC_KEYMAP_EEPROM_MAX_ADDR 2047
+#define DYNAMIC_KEYMAP_LAYER_COUNT 3
\ No newline at end of file
```

**Comment by tzarc** [comment]
@ramonimbao can you confirm whether you're okay dropping the number of layers on `wete` to 3?
If it's a 'yes', @xyzz will be able to mod the PR and we can hopefully push things through.

**Comment by tzarc** [comment]
Seems we've picked up `durgod/k320` since I last looked. @dkjer can we decrease the layer count in the same way? This wear leveling PR can't go in unless the defaults are currently honoured -- work can be done afterwards to reinstate the available space increase.

**Comment by ramonimbao** [comment]
> can you confirm whether you're okay dropping the number of layers on `wete` to 3?

Yep, please do. Thanks

**Comment by dkjer** [comment]
> can we decrease the layer count in the same way?

@tzarc Yes, that is fine

**Comment by tzarc** [comment]
@ramonimbao @dkjer thanks guys, much appreciated.

**Comment by xyzz** [comment]
ok, updated

**Review by tzarc — APPROVED** [review]

File: keyboards/ramonimbao/wete/keymaps/via/config.h [hunk 1]
```diff
@@ -1 +1,2 @@
-#define DYNAMIC_KEYMAP_EEPROM_MAX_ADDR 2047
+#define DYNAMIC_KEYMAP_LAYER_COUNT 3
```
Comment (zvecr) [inline keyboards/ramonimbao/wete/keymaps/via/config.h:1]:
```suggestion
#pragma once

#define DYNAMIC_KEYMAP_LAYER_COUNT 3
```

File: keyboards/durgod/k320/keymaps/via/config.h [hunk 2]
```diff
@@ -0,0 +1 @@
+#define DYNAMIC_KEYMAP_LAYER_COUNT 3
```
Comment (zvecr) [inline keyboards/durgod/k320/keymaps/via/config.h:1]:
```suggestion
#pragma once

#define DYNAMIC_KEYMAP_LAYER_COUNT 3
```

**Review by zvecr — CHANGES_REQUESTED** [review]

**Comment by tzarc** [comment]
Asking again..... @unxmaal -- can you confirm that you don't own any of the following keyboards: `akegata_denki/device_one`, `chavdai40/rev1`, `chavdai40/rev2`, `peiorisboards/ixora`, `polilla/rev1`, or `vinta`?

**Comment by unxmaal** [comment]
> Asking again..... @unxmaal -- can you confirm that you don't own any of the following keyboards: `akegata_denki/device_one`, `chavdai40/rev1`, `chavdai40/rev2`, `peiorisboards/ixora`, `polilla/rev1`, or `vinta`?

Confirmed: I've never heard of these boards. I have a H87a, DZ60, and a New Model F. Maybe I need more boards?

**Comment by tzarc** [comment]
> Confirmed: I've never heard of these boards. I have a H87a, DZ60, and a New Model F. Maybe I need more boards?

We all do, mate! Cheers for the confirmation :)

**Comment by dkjer** [comment]
@mcoirault Heads up that the K320 VIA keymap will temporarily be limited to 3 layers

File: tmk_core/common/chibios/eeprom_stm32.h [hunk 3]
```diff
@@ -69,16 +69,23 @@
...
+    "flash memory for emulated eeprom is too small; for correct functionality ensure it is at least 8x FEE_DENSITY_BYTES");
+
+#if defined(DYNAMIC_KEYMAP_EEPROM_MAX_ADDR) && (DYNAMIC_KEYMAP_EEPROM_MAX_ADDR >= FEE_DENSITY_BYTES)
+#error emulated eeprom: DYNAMIC_KEYMAP_EEPROM_MAX_ADDR is greater than the FEE_DENSITY_BYTES available
+#endif
+
+void     EEPROM_Init(void);
 void     EEPROM_Erase(void);
-uint16_t EEPROM_WriteDataByte(uint16_t Address, uint8_t DataByte);
+void     EEPROM_WriteDataByte(uint16_t Address, uint8_t DataByte);
```
Comment (dkjer) [inline tmk_core/common/chibios/eeprom_stm32.h:90]: Why are we no longer returning FLASH_Status, here?  It may be useful to know if a write failed, and why

Comment (xyzz) [reply tmk_core/common/chibios/eeprom_stm32.h:90]: the return code was never used, plus it is unclear what should be done in an event of an error

**Comment by dkjer** [comment]
@xyzz Have you already started a follow-up branch for increasing the available simulated EEPROM beyond 1024?  I have some ideas around that I'd like to discuss, perhaps in discord?

**Comment by xyzz** [comment]
I have not, feel free to ping me on qmk discord to discuss though.

**Pull request closed without merging**